ENH: pandas 0.18.0 support

jairideout opened the pull request:
Fixes #1308. scikit-bio now depends on pandas >= 0.18.0.

Added support for `pandas.RangeIndex`, lowering the memory footprint of default integer index objects. `Sequence.positional_metadata` and `TabularMSA.positional_metadata` now use `pd.RangeIndex` as the positional metadata index. `TabularMSA` now uses `pd.RangeIndex` as the default index.  Usage of `pd.RangeIndex` over the previous `pd.Int64Index` [should be transparent](http://pandas.pydata.org/pandas-docs/version/0.18.0/whatsnew.html#range-index), so these changes should be non-breaking to users.

Added `reset_index=False` parameter to `TabularMSA.append` and `TabularMSA.extend` for resetting the MSA's index to the default index after appending/extending. `TabularMSA.append` and `TabularMSA.extend` now require one of `minter`, `index`, or `reset_index` to be provided when incorporating new sequences into an MSA. Previous behavior was to auto-increment the index labels if `minter` and `index` weren't provided and the MSA had a default integer index, otherwise error.

Fixed doctests to work with pandas 0.18.0 ("integers" stored in a float column now repr with a trailing `.0`).

jairideout commented on skbio/alignment/tests/test_tabular_msa.py:
```diff
@@ -1793,79 +1884,75 @@ def test_index_override(self):
 
 
 class TestAppend(unittest.TestCase):
```
I re-wrote/organized/named the tests for `append` and `extend`, the diff here won't be very helpful to reviewers.

jairideout commented:
Travis isn't seeing this PR at all, going to close and reopen.

jairideout commented:
Tests passed, ready for review

ebolyen commented:
Sorry for the delay! Reviewing now.

ebolyen commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -1703,18 +1720,20 @@ def append(self, sequence, minter=None, index=None):
-            If neither `minter` nor `index` are provided and the MSA has a
-            non-default index.
+            If exactly one choice of `minter`, `index`, or `reset_index` is not
+            provided.
```
:+1:

ebolyen commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -1847,59 +1892,89 @@ def extend(self, sequences, minter=None, index=None):
 
         """
-        if minter is not None and index is not None:
+        if sum([minter is not None, index is not None, reset_index]) != 1:
```
Use `any` here instead of `sum`

ebolyen replied:
oh, never mind. this error is catching a couple cases.

ebolyen replied:
use `bool(reset_index)` as `sum` will act weird in a truthy context. Assuming with allow `reset_index` to ducktype its boolean-ness

jairideout replied:
Good catch!

jairideout replied:
Fixed and added a couple test cases

ebolyen commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -2117,8 +2192,8 @@ def join(self, other, how='strict'):
-        0    42     1  NaN
-        1    43     2  NaN
+        0  42.0     1  NaN
+        1  43.0     2  NaN
```
lovely...

ebolyen commented on skbio/alignment/tests/test_tabular_msa.py:
```diff
@@ -562,7 +630,7 @@ def test_reassign_index_with_mapping_callable(self):
 
         msa = TabularMSA(seqs, index=[0, 1, 2])
-        msa.reassign_index(mapping=str)
+        msa.reassign_index(mapping=lambda e: str(e))
```
Why the η-expansion?

jairideout replied:
This was a backwards-incompatible change to `pd.Series.rename`. Passing `mapping=str` now modifies the series' `name` attribute (feature? bug? [the docs](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.rename.html) aren't helpful).

Do you think we should continue using `Series.rename` as a passthrough for `mapping`, or roll our own?

jairideout replied:
After offline discussion with @ebolyen, we're going to roll our own! Using `Series.rename` as a passthrough has some other issues, like allowing scalars and list-like to rename the Series' name.

ebolyen commented:
@jairideout this was a doozy! Only a few questions, otherwise looks good!

jairideout commented:
Thanks for reviewing @ebolyen, sorry this was a painful one. I addressed your comments and had [a question](https://github.com/biocore/scikit-bio/pull/1325#discussion_r59898562), otherwise this should be good to go 🎸

jairideout commented:
@ebolyen ping on this one

jairideout commented:
@ebolyen ready for final review/merge

ebolyen commented on skbio/alignment/tests/test_tabular_msa.py:
```diff
@@ -630,10 +640,14 @@ def test_reassign_index_with_mapping_callable(self):
 
         msa = TabularMSA(seqs, index=[0, 1, 2])
-        msa.reassign_index(mapping=lambda e: str(e))
+        msa.reassign_index(mapping=str)
```
:+1:

ebolyen commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -1701,7 +1701,15 @@ def reassign_index(self, mapping=None, minter=None):
+                              for label in self.index]
+            elif callable(mapping):
+                self.index = [mapping(label) for label in self.index]
+            else:
```
This turned out very simple...

Now isn't there an edge case with respect to `tuples` and `pd.MultiIndex`? I don't remember if assignment always converted to a MultiIndex where possible. I'll double check that.

ebolyen replied:
Looks like it does, awesome!

ebolyen commented:
Thanks @jairideout!

wasade commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -1701,7 +1701,15 @@ def reassign_index(self, mapping=None, minter=None):
         if mapping is not None:
-            self._seqs.rename(mapping, inplace=True)
+            if isinstance(mapping, dict):
+                self.index = [mapping[label] if label in mapping else label
```
I know this is merged already, but just took a peak. Just out of curiosity, is the ternary form favored over `mapping.get(label, label)` or does this fall to developer preference?

ebolyen replied:
Developer preference I think. My preference probably matches the ternary as it looks a little "prettier" and it's purpose feels clearer in this context. But there's nothing wrong with your approach either.

wasade replied:
Okay, awesome. Thank you. I love the ternary
On May 4, 2016 1:17 PM, "Evan Bolyen" <notifications@github.com> wrote:

> In skbio/alignment/_tabular_msa.py
> <https://github.com/biocore/scikit-bio/pull/1325#discussion_r62106829>:
>
> > @@ -1701,7 +1701,15 @@ def reassign_index(self, mapping=None, minter=None):
> >                  "Cannot use both `mapping` and `minter` at the same time.")
> >
> >          if mapping is not None:
> > -            self._seqs.rename(mapping, inplace=True)
> > +            if isinstance(mapping, dict):
> > +                self.index = [mapping[label] if label in mapping else label
>
> Developer preference I think. My preference probably matches the ternary
> as it looks a little "prettier" and it's purpose feels clearer in this
> context. But there's nothing wrong with your approach either.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1325/files/0183591e24cca3243bebd454fb76ef2519933c58..a988e7faf40030a92e90dac3f4f0518e10ea2849#r62106829>
>

jairideout commented:
Awesome, thanks @ebolyen!
